_posts/2024-06-14-remote-code-execution-engine.md

```diff
@@ -0,0 +1,1426 @@
+addSbtPlugin("com.eed3si9n" % "sbt-assembly" % "2.1.1") // SBT plugin for using assembly command
+```
+
+## 3. Project Architecture
+
+After a few iterations I came up with the architecture that can be horizontally scaled, if required. 
+Ideally, such projects must be scaled easily as long as the load is increased. 
+
+For that we use tools such as Kubernetes or other container orchestration platforms. To make local development and deployment simpler we'll be using docker containers. More precisely we'll be using `docker-compose` to run a few containers together so that they form the cluster.
+
+`docker-compose` doesn't support scalability out of the box because it's static, it means that we can't magically add new `worker` node to the running system. Again, for that we'd use Kubernetes, but it is out of the scope of this project.
+
+We have a `master` node and its role is to be the task distributor among `worker` nodes. 
+
+`http` is exposed on `master` node, acting as a gateway to outside world.
+
+### 3.1 Architecture Diagram
+
+![alt "Project skeleton"](../images/braindrill/diagram.png)
+
+Let's explain step by step what happens during code execution:
+
+- User sends request at POST `http://localhost/python` with `python` code attached to request body
+- `http` processes the request
+- randomly chosen `load balancer` sends message to `worker router` on some node
+- `worker router` sends message to `worker` which starts execution
+- ... skipping lots of details ...
+- `worker` responds to `load balancer` once the code is executed
+- `load balancer` forwards message to `http` and user receives the output
+
+### 3.2 Docker runtime
+
+There are a few options to my knowledge for executing the submitted code with the help of docker containers. All I can do is redirect you to the great [StackOverflow post](https://stackoverflow.com/questions/27879713/is-it-ok-to-run-docker-from-inside-docker) which expands on this issue.
+
+Long story short, I decided to go with [DooD - Docker out of Docker](https://tdongsi.github.io/blog/2017/04/23/docker-out-of-docker/).
+
+`DooD` Uses the host's Docker daemon by mounting `/var/run/docker.sock` into the container.
+
+It means that if the host machine has the docker engine installed on it:
+- it can start container `A` using `/var/run/docker.sock` which listens for commands
+- it can start "sibling" container `B` by executing commands such as `docker run ...` from container `A`, using `/var/run/docker.sock`
+
+In our domain terms:
+- container `A` will be `worker-1` container, which will be starting sibling containers
+- sibling containers will be short-lived, limited (CPU, RAM, timeout) containers which run `python`, `java` and such processes
+
+Advantages of `DooD`:
+- Efficient: Direct use of host resources.
+- Simple: No extra Docker daemons.
+- Fast: No nested virtualization overhead.
+- Resource-efficient: Shared daemon optimizes resource use.
+
+Disadvantages:
+- Security risk: Access to host Docker daemon.
```

Comment:
Maybe overhead starting/stopping dockers all the time?

Reply:
Yeah, will add that